### PR TITLE
Detach wallet

### LINKED ISSUE #260: [Blockchain data fetching] Proposal: Expect external service to notify daemon about wallet activites
## TL;DR

tdex-daemon should focus entirely on Atomic Swap & Trading, not on solving blockchain wallet indexing.

## Intro

We have made an effort early on to depends on Electrs REST API as a way to fetch blockchain data needed by the trade activities of the daemon (unspents, spents, transactions).

The initial idea was to privilege the easy of use and portability: ie. I just need to switch an endpoint from a public instance to a personal one connect to my own elements node.
In the end this revealed wrong. An Electrs instance is resource hungry (>16GB RAM) resulting no-one will self-host it and use his own Elements node in the end, making the project somewhat less decentralized and degrading the privacy of market makers depending on public instances. Moreover public instances will likely rate-limit or ban IPs that makes too much requests.

## Crawler

The crawler component is basically using HTTP polling to fetch data about addresses and transactions we are interested. There are couple of improvements on the line that will lower the amount of requests per second we make to external service https://github.com/TDex-network/tdex-daemon/pull/259  and the additional feature of using **Elements as poor man explorer** In this PR https://github.com/TDex-network/tdex-daemon/pull/255 can help privacy conscious users and improve overall decentralization of the solution.

## ZMQ support

Elements node, as Bitcoin Core, supports pushing events via ZMQ channels, making suitable as a way for using Elements as block/tx validators, leaving other applications to consume the blocks and scan it for know addresses and transactions. 

## Wallet tracker 

I have mad a PoC here https://github.com/tiero/ewt that does this, subscribe to ZMQ events, register his own array of scripts of the tx Im interested, filter the raw tx and "fire a notification". 

This would sits between the elements node and the daemon (or any other applications really). In the same way, the crawler using the classic HTTP polling will do the same thing (as a standalone daemon) using either Electrs REST API or a tweaked Elements node as poor man explorer https://github.com/TDex-network/tdex-daemon/pull/259#discussion_r582786551

## Reactive model

My proposal is to better separate concerns and allow different teams/solutions to be plug-in, standardizing a single way to "notify" the tdex-daemon of when a new address has got new unspent or a txid has been included in a block, either using a Pub/Sub or bidirectional streaming interface

The flow would be:

1) tdex-daemon "register" an xpub or list of addresses in the external **wallet tracker**
2) wallet tracker detects "wallet activity" like (un)spents or txid get included in block, filtering raw blocks/tx
3) wallet tracker notify the daemon of the new event and attach blockchain data 

In case of restore of HD wallet, to fetch data of the past we would still need to relay on unary calls to wither the Elements node or Explorer to discover unused addresses (ie. gap limit) but this is not of a concern since is one-time cost at the beginning.

## Streaming

The bi-directional streaming is an appealing solution since we do not need to know the external interface where we need to subscribe and other software/explorer/wallet tracker need to just call HTTP2 endpoint to start the bi-directional channel

if coupled with gRPC an ideal API would be (just a quick sketch, not well tought)

```proto

service WalletTracker {
    rpc Notify(stream Filter) returns (stream Data) {}
}

message Filter {
    repeated string scripts = 1
}

message Data {
    enum Type {
        OUTPOINT = 0;
        TX = 1;
    }
    Type type = 1;
    // whatever message we want to attach ie. amount, value & asset for unspent
    bytes data = 2;
}

```

### LINKED ISSUE #309: [Badger] split each Repository in his own datastore on the filesystem

### LINKED ISSUE #413: Blockchain data sourcing SPV/Neutrino: connect to p2p network to receive relevant block that contains the scripts we are interested to
At the moment, having the unspents cached in-memory only, means that every update/restart, will take lot of time to re-sync the unspents of the wallet, since we need to make network calls for each address we discover.

The definitive approach is to just use the least dependency from the explorer (ie. GET /block) to fetch the chain of blocks, parse them and cache relevant unspents.

`btwallet`  has a `chain` package that could be adapted to support many node backends such as Electrs explorer or even an Elements node via authenticated JSON RPC.
https://github.com/btcsuite/btcwallet/tree/master/chain

### LINKED ISSUE #415: Refactor multiplexing to make it work with reverse proxy such as Caddy
Ref: https://github.com/caddyserver/caddy/issues/3997

This way we could serve behind caddy server reverse proxy and automatic TLS management easily

Code can be taken from here https://github.com/tiero/zion/blob/serverless/cmd/zion/main.go#L95

### LINKED ISSUE #482: Detach Signer to be able to plug external key management components
Let's say we want to have GDK to handle keys and signing or having KMS/Lambda doing this externally etc.. would be better to detach keys from the daemon so we can plug many "signing backend" without embedding in the main binary

### LINKED ISSUE #539: [GetFeeFragmenterAddress] Sometimes returns `Transaction Conflict. Please retry`
It happens from UI only. 
Among other things I call `GetFeeAddress` at the same time when mounting the React component, and it might be the cause of the issue.

```
export const FeeDeposit = (): JSX.Element => {
  const { refetch: refetchGetFeeBalance } = useGetFeeBalanceQuery();
  const { data: feeFragmenterAddress, refetch: refetchFeeFragmenterAddress } =
    useGetFeeFragmenterAddressQuery({ numOfAddresses: 1 });
  const { data: feeAddress, refetch: refetchFeeAddress } = useGetFeeAddressQuery();
  const [feeFragmenterSplitFunds] = useFeeFragmenterSplitFundsMutation();
  const [claimFeeDeposits] = useClaimFeeDepositsMutation();
```

### LINKED ISSUE #552: DropMarket: do not reuse the account index of a dropped market
In case of dropping a market, after having some transaction on it, we must prevent re-using it for different asset pair.

In case of restore, we cannot understand which asset pair was the last automatically (or in a very opininated/complex way).

Simple solution will be to restore it and having there "reserved" and his account index not usable

### LINKED ISSUE #654: Allow user to decide which market to be "fixed" when restoring markets
## Why

The end goal is to be able to allow the operator to totally control his TDEX daemon instance entirely from RPC (therefore GUIs, dashboards) without having to mess with SSH command-line. This allows separating the duty of WHO is packaging & deploying the instance and WHO instead is the market maker using it via clients.

## What 
Currently, the restore scans for all addresses and uses some constants such as `TDEX_BASE_ASSET` or `TDEX_QUOTE_ASSET` to know which asset to be fixed for market discovery.

### Proposal 1 
I propose to pass wither of the two variables ALSO via RPC when user is invoking the `InitWallet` RPC with `restore=true`.

This will make sure to override via RPC whatever is set (or not set) via config/env

### Propoasl 2 

Persist and allow to set/unset the entirety of the `config` via RPC. Maybe this is more future-proof for other things we want to expose to operators' UIs?

### PR DESCRIPTION
**Pull request by altafan**
This contains a HUGE refactor of the daemon with the main purpose of detaching the embedded wallet and make it connect to an external ocean wallet instead. Secondary, this refactors the scaffolding and file naming for app and domain layers in order to increase the readability and maintainability of the codebase.

## Changes

* All stuff related to Vault and Unspent domain/repo has been dropped.
* A new portable interface ~`OceanWallet`~ `WalletService` has been introduced.
* Strong refactor of how the grpc services are started to decrease overall code complexity
* Refactor of the env vars to configure the daemon at startup
* Updates to external pkgs in order to user latest go-elements and psetv2.
* New tdex-daemon/v2 proto that requires updates to clients such as the feeder or the dashboard that interact with operator interface.
* Updated e2e test in order to use dockerized services instead of compiled binaries.

Note: I'll provide more detailed clues about the macro changes listed above in later comments.

Closes #260.
Closes #309.
Closes #413.
Closes #415.
Closes #482.
Closes #539.
Closes #552.
Closes #654. (Removed env vars to fix asset)

Please @tiero review.

### CODE REVIEW

File: internal/core/application/operator/fee_fragmenter_account.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+	chRes <- fragmenterReply{
+		fmt.Sprintf(
+			"fetched funds with total LBTC balance of %d, to be split into %d "+
+				"fragments of %d sats each", balance, numFragments, feeFragmentAmount,
+		), nil,
+	}
+
+	// make sure the fee account is created.
+	s.wallet.Account().CreateAccount(ctx, domain.FeeAccount)
```
Comment (sekulicd) [inline internal/core/application/operator/fee_fragmenter_account.go:106]: I see in many places errors were not checked, this looks problematic even if we are sure error wont happen or it is not important, for eg. here we are calling other process(ocean) and in case ocean is not reachable error will be "silent" and we wont know.
To find all places where error is not handled use golangci-lint run --disable-all -E errcheck

File: internal/core/ports/ocean_wallet.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
+package ports
+
+import "context"
+
+type OceanWallet interface {
```
Comment (sekulicd) [inline internal/core/ports/ocean_wallet.go:5]: In general core/ports should not be aware about its implementation, in this case ocean, as tdexd doesnt care about it all it cares is that it is using Wallet interfaces with all its methods, so i would rename file to wallet_service or smtn.
Also i think we should decide to use one name for types/files and instead service/manager we should pick one, my favourite is Service, wtf is Manager :)
So in this case I would name file wallet_service, main interface would be WalletService and other intf would be Wallet, Account, Transaction etc

File: internal/config/config.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	minPercentageFee = 0.01
+	maxPercentageFee = float64(99.99)
+
+	httpsProtocol = "https"
+)
+
+var vip *viper.Viper
+var defaultDatadir = btcutil.AppDataDir("tdex-daemon", false)
+
+func init() {
```
Comment (sekulicd) [inline internal/config/config.go:84]: Personally, if i am not forced, I am done using init()  :)
I think it only inserts unpredictable/uncontrolled behaviour

File: internal/core/application/wallet/service.go [hunk 4]
```diff
@@ -0,0 +1,374 @@
+package wallet
```
Comment (sekulicd) [inline internal/core/application/wallet/service.go:1]: If i am not wrong your intention/motivation was to make trade/operator service more readable/maintainable by splitting it to more files and adding folders, I like that but i think we could have keep all it in same folder and not introduce new packages like operator, pubsub, trade, wallet(we should have one pkg for application layer 'application)
After detaching wallet our methods are no longer so long and could have been kept in one file or atleast in same folder and if needed some method could be separated in another file eg. operator_service_fee_fragmenter_split_fund.go for FeeFragmenterSplitFunds.

File: internal/core/domain/market_actions.go [hunk 5]
```diff
@@ -2,120 +2,68 @@ package domain
```
Comment (sekulicd) [inline internal/core/domain/market_actions.go:2]: market_action.go file name looks untypical, i would keep it in market.go

**Review by sekulicd — COMMENTED** [review]
In general its great we detached wallet, it will be much easier to maintain.
Good job @altafan :)

**Comment by sekulicd** [comment]
Also maybe we should have docker_compose in root and make dev in Makefile which would start tdexd and ocean

File: internal/interfaces/grpc/service_one_port.go [hunk 6]
```diff
@@ -87,372 +151,238 @@ func NewServiceOnePort(opts ServiceOptsOnePort) (interfaces.Service, error) {
...
 
-	address := s.opts.Address
-	grpcServer := grpc.NewServer(
-		unaryInterceptor, streamInterceptor,
-	)
-	walletUnlockerHandler := grpchandler.NewWalletUnlockerHandler(
-		s.opts.WalletUnlockerSvc, adminMacaroonPath,
+	grpcServer := grpc.NewServer(serverOpts...)
+	walletHandler := grpchandler.NewWalletHandler(
+		s.opts.AppConfig.WalletService().Wallet(), s.opts.BuildData,
```
[the remark this answers is not in the document]
Comment (altafan) [reply internal/interfaces/grpc/service_one_port.go:280]: `WalletService` is actually an application service like operator or trade ones.

File: internal/config/config.go (hunk 3 above)
Comment (altafan) [reply internal/config/config.go:84]: will move to InitConfig func

File: internal/core/application/wallet/service.go (hunk 4 above)
Comment (altafan) [reply internal/core/application/wallet/service.go:1]: Consider that all app services' factories and interfaces are in `application` package, so the upper layers aren't actually aware of their implementations, and all these "internal" services become just implementation details, like an implementation of the repo manager is for an app service to make a comparison.

**Comment by altafan** [comment]
@sekulicd do you have any suggestion for the CodeQL failing action (`InsecureSkipVerify`)? I tried several things but none of them seemed to work. Not sure why now it's not working now. I didn't actually change that and before this PR it was not giving problems...

File: internal/core/application/mocks_test.go [hunk 7]
```diff
@@ -1,648 +0,0 @@
-package application_test
```
[the remark this answers is not in the document]
Comment (altafan) [reply internal/core/application/mocks_test.go:1]: I didin't know about this tool! Thanks for sharing :)

File: internal/infrastructure/market-strategy/pluggable/strategy.go [hunk 8]
```diff
@@ -1,7 +1,8 @@
-package domain
```
[the remark this answers is not in the document]
Comment (altafan) [reply internal/infrastructure/market-strategy/pluggable/strategy.go:1]: Maybe you're referring to pluggable strategy, not swap parser?

I actually moved this strategy here because it was in core/domain/ originally. I agree with you that since this doesn't depend on domain it can be moved in pkg. I'm moving it to pkg/marketmaking.

[the remark this answers is not in the document]
Comment (altafan) [reply internal/infrastructure/market-strategy/pluggable/strategy.go:1]: for what concerns swap parser, instead, since it depends on domain, it's correct to have it in infrastructure, since it used proto stubs which are are a thing of our infra and also internal stuff like the Swap domain.

File: internal/interfaces/grpc/handler/operator.go [hunk 9]
```diff
@@ -987,511 +787,180 @@ func (o operatorHandler) marketFragmenterSplitFunds(
...
-				Index: int32(u.Outpoint.Index),
-			},
-			Value: u.Value,
-			Asset: u.Asset,
-		})
-	}
-	return res
+	return &daemonv2.ListDepositsResponse{
+		AccountName: accountName,
+		Deposits:    depositsInfo(deposits).toProto(),
```
[the remark this answers is not in the document]
Comment (altafan) [reply internal/interfaces/grpc/handler/operator.go:942]: The fact is that in core/ports we have only definitions of interface types and this makes them perfect to be used as cross-layer types, instead of having repetition of types and validation functions across layers.

> So ideally each layer should have types.go file where each type would be defined validated and where there would be conversion to layer bellow.

But I agree with you on this, and I just noticed that there are no validation checks in the app services at the moment. I'm going to add them. I have this idea of adding the validation functions in core/ports so every layer can just make use of them instead of implementing their own methods.

What do you think of this approach? it seems it solves both the problems you stated and the one i wanted to solve which was preventing to have basically lot of duplicated code in all layers.

File: internal/interfaces/grpc/service_one_port.go [hunk 10]
```diff
@@ -87,372 +151,238 @@ func NewServiceOnePort(opts ServiceOptsOnePort) (interfaces.Service, error) {
...
+		adminMacaroonPath,
+		s.onInit, s.onUnlock, s.onLock, s.onChangePwd,
```
[the remark this answers is not in the document]
Comment (altafan) [reply internal/interfaces/grpc/service_one_port.go:282]: Doesn't seem so anti-pattern actually https://stackoverflow.com/questions/25203988/go-and-callbacks and looks cleaner for this very specific purpose since we can handle the unlocking of macaroon store in a sync way.

File: internal/infrastructure/market-strategy/pluggable/strategy.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (altafan) [reply internal/infrastructure/market-strategy/pluggable/strategy.go]: done.

File: internal/core/application/mocks_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (altafan) [reply internal/core/application/mocks_test.go:1]: I played with the tool a little bit and my preferred solution is to create a mocks directory in the folder layer (`domain` in this case), then import the mocks from there when testing. The resulting command (ran after cd-ing into internal/core/domain) is: `mockery --name=SwapParser --recursive --structname=MockSwapParser --filename=swap.go`. The filename should be the same of the one containing the interface being mocked.

**Comment by tiero** [comment]
> CodeQL

Shouldn't //nosec enough?

File: internal/core/application/operator/market_fragmenter_account.go [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+			amount := percentageAmount(quoteBalance, percentage)
+			for i := 0; i < n; i++ {
+				addr := addresses[offset+i]
+				outputs = append(outputs, output{market.QuoteAsset, amount, addr})
+			}
+		}
+	}
+
+	chRes <- fragmenterReply{
+		"crafting and broadcasting transaciton to send funds to market account",
```
[the remark this answers is not in the document]
Comment (altafan) [reply internal/core/application/operator/market_fragmenter_account.go:200]: done.

File: internal/core/application/operator/fee_fragmenter_account.go (hunk 1 above)
Comment (altafan) [reply internal/core/application/operator/fee_fragmenter_account.go:106]: done.

File: internal/core/ports/ocean_wallet.go (hunk 2 above)
Comment (altafan) [reply internal/core/ports/ocean_wallet.go:5]: done.

File: internal/config/config.go (hunk 3 above)
Comment (altafan) [reply internal/config/config.go:84]: done.

**Comment by altafan** [comment]
I solved by manually dismissing those alerts with "Won't fix" reason.

**Comment by altafan** [comment]
## Changes to env vars

* **Removed:**
     *  `TRADE_EXPIRY_TIME` - the expiration time is given by the ocean wallet when selecting the utxos. At the moment ocean protos don't expect this value to be set within the request message. Until we eventually commit this change in ocean, this env var won't be restored.
     * `EXPLORER_*` - The daemon doesn't make use of an explorer service anymore.
     * `BASE_ASSET`, `QUOTE_ASSET` - The daemon doesn't make use of fixed base/quote asset anymore. Any kind of market can be created.
     * `NETWORK`, `NATIVE_ASSET` - The daemon retrieves info about the network from the underlying ocean wallet.
     * `CRAWL_INTERVAL` - The daemon doesn't have any underlying component taking care of watching the blockchain anymore.
     * `TRADE_SATS_PER_BYTE` - Always defaults to 0.1 sats/VByte. This env var was removed for easiness, but it can (should?) be restored.
     * `MAX_FAILING_REQUESTS`, `FAILING_RATIO` - The daemon doesn't make use of any ratelimiter/circuitbreaker anymore.
     * `RESCAN_*` - Wallet restoration isn't yet available since it's a feature under development, but in general maybe better to provide these info in the request message.
* **Renamed** (this refactor was the opportunity to give some env vars a more expressive name): 
     * `DATA_DIR_PATH` --> `DATADIR`
     * `DEFAULT_FEE` --> `PERCENTAGE_FEE`
     * `SSL_KEY`, `SSL_CERT` --> `TRADE_TLS_KEY`, `TRADE_TLS_CERT`
* **Added:**
     * `WALLET_ADDR` - address of the ocean wallet service to connect to
     * `DB_TYPE` - type of db to use (for now only `badger`, soon we will support also `postgres`)

**Comment by tiero** [comment]
> The daemon doesn't make use of fixed base/quote asset anymore. Any kind of market can be created.

how do we account for it when restoring? how do we know which is base and which is quote?

**Comment by tiero** [comment]
>  This env var was removed for easiness, but it can (should?) be restored.

Maybe this should be calculated dynamically looking at meempool? user may want to turn off this as "priority" or not

File: internal/core/application/mocks_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sekulicd) [reply internal/core/application/mocks_test.go:1]: That sounds ok.
Maybe u can add make cmd for generating mocks like i did in [here](https://github.com/XDEX-SA/xdex-price-alerts/blob/b7da215d3d2bb36e610a62d8198828c2a54c5dae/Makefile#L99) since interface can change and then u can easily recreate all mocks with one command.

File: internal/infrastructure/market-strategy/pluggable/strategy.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sekulicd) [reply internal/infrastructure/market-strategy/pluggable/strategy.go]: I would define infrastructure as another process inside our deployment(DB etc.)

File: internal/interfaces/grpc/service_one_port.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sekulicd) [reply internal/interfaces/grpc/service_one_port.go:285]: Yeah, but here u are not passing WalletService but its internal interface -> Wallet() -> ports.Wallet

File: internal/interfaces/grpc/handler/operator.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sekulicd) [reply internal/interfaces/grpc/handler/operator.go:940]: I think the cleanest(and yeah most boring way) would be to have types for each layer.

**Comment by altafan** [comment]
> how do we account for it when restoring? how do we know which is base and which is quote?

The user could pass a list of base assets when restoring a wallet. In general, my intention was to move this kind of config from env var to request message so the user can pass this info at runtime.

**Comment by altafan** [comment]
> Maybe this should be calculated dynamically looking at meempool? user may want to turn off this as "priority" or not

Let's put this on pause for now. We'll create a dedicated ticket as soon as this PR gets merged so we can discuss about it further

File: internal/infrastructure/market-strategy/pluggable/strategy.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (altafan) [reply internal/infrastructure/market-strategy/pluggable/strategy.go]: you're not that wrong actually :).. so shall I move the swap-parser to application? we can't have it in pkg because it depends on domain which is an internal layer. Application seem the right package unless we create a another folder inside/outside core. What do you think? @sekulicd

File: internal/interfaces/grpc/handler/operator.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (altafan) [reply internal/interfaces/grpc/handler/operator.go:940]: What I don't really like about this pattern is that we end up having lots of code repetition, especially for validation functions, and this can easily introduce bugs and, in general, makes the codebase harder to maintain too.

On the other side, expecting to have only interface types in ports make them usable as cross-layer types. We end up having validation functions only on higher layers (interfaces), where we want to validate what the user is passing in the request message, while on lower ones we use only internal concrete types, where we are sure to use only correct values.

File: internal/infrastructure/market-strategy/pluggable/strategy.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sekulicd) [reply internal/infrastructure/market-strategy/pluggable/strategy.go]: Yes

[the remark this answers is not in the document]
Comment (altafan) [reply internal/infrastructure/market-strategy/pluggable/strategy.go]: Done.

File: internal/core/application/mocks_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (altafan) [reply internal/core/application/mocks_test.go:1]: Done.

**Comment by altafan** [comment]
## Changes to tdex-daemon/v2 proto compared to v1:

* Removed
   * `WalletService` - this service was meant to be used as operator's personal wallet. He can now use directly ocean for this purpose.
   * `OperatorService/Claim*` rpc - not needed anymore because the daemon gets notified about deposits/withdrawals from the ocean wallet. This is a breaking change and requires clients making use of such apis (dashboard) to be updated accordingly.
   * `OperatorService/GetMarketBalance` rpc - removed since duplicate of GetMarketInfo.
   * `OperatorService/GetMarketCollectedFees` rpc - merged into GetMarketReport.
* Renamed:
   * `WalletUnlockerService` --> `WalletService`
   * `WalletUnlockerService/IsReady` --> `WalletService/GetStatus`
   * `OperatorService/GetInfo` --> `WalletService/GetInfo`
   * `OperatorService/Get*Address` --> `OperatorService/Get*Addresses` - moved to plural since the rpc returns a list of newly derived addresses.
* Added:
   * `WalletService/LockWallet` rpc - useful to lock the daemon and its wallet without switching it off. This action disables every interface but the `WalletService` one like it was just started and waiting to be unlocked.
* Other changes:
   * Replaced all `bytes` message fields with `string`
   * Fixed all enums with `*_UNSPECIFIED = 0` first value as per buf recommended practice.
   * Expect a list of `Receivers` in all Withdrawal rpcs to make them consistent.
   * Use `int64` for all timestamp message fields.
   * Balance returned by OperatorService/GetMarketInfo is expressed as map of pairs `<asset_hash, { unconfirmed_balance, confirmed_balance, locked_balance, total_balance }>`

**Comment by tiero** [comment]
> Changes to tdex-daemon/v2 proto compared to v1:

cc/ @Janaka-Steph

File: internal/interfaces/grpc/service_one_port.go [hunk 12]
```diff
@@ -339,7 +340,24 @@ func (s *serviceOnePort) newServer(
 		grpcweb.WithOriginFunc(func(origin string) bool { return true }),
 	)
 
-	handler := router(grpcServer, grpcWebServer, grpcGateway)
+	tdexConnectSvc, err := httpinterface.NewTdexConnectService(
+		s.opts.AppConfig.WalletService().Wallet(),
+		s.validatePassword,
+		adminMacaroonPath,
```
Comment (tiero) [inline internal/interfaces/grpc/service_one_port.go:346]: why do we need macaroon passed in the tdexconnect service?

**Comment by Janaka-Steph** [comment]
- Getting `GetMarketReport failure - RpcError: not implemented`, method seems missing.

- Getting `ListTrades failure - RpcError: Skip must be set to a positive number` if `page: { number: 0, size: 5 }`. Number should be at least 1. What should be the behaviour of `number: 0`? Error message should be more explicit.

File: internal/interfaces/grpc/service_one_port.go (hunk 12 above)
Comment (altafan) [reply internal/interfaces/grpc/service_one_port.go:346]: i didn't change this, it's inherited from master https://github.com/tdex-network/tdex-daemon/blob/master/internal/interfaces/grpc/service.go#L357-L364

**Comment by sekulicd** [comment]
I want to re-discuss/challenge the way we are handling types in different layers in this PR.

Purpose of ports pkg is to isolate core from calls to 3rd party apps/infra.
In this PR we moved almost all types to ports, even some that doesnt "leave" our core.
I will use GetMarketReport method as an example and its arg TimeRange type.
TimeRange is simple struct or DTO used only to pass some data around as a whole, as opposed to several individual variables.
Before this TimeRange was defined as [struct](https://github.com/tdex-network/tdex-daemon/blob/fe2fee347287fc093a04108bd20d19b3b11dd8b0/internal/core/application/types.go#L513) with its validation/toDate methods and now it is defined in ports as an [interface](https://github.com/altafan/tdex-daemon/blob/475d7548fc16b110777f6fb6633bd09b7a7518af/internal/core/ports/types.go#L151).
Representing [DTO](https://herbertograca.com/2020/06/23/dto-data-transfer-objects/)(in this case TimeRange) as an interface, introduces only unnecessary complexity.
In this PR we are doing validation in upper layer([handler](https://github.com/altafan/tdex-daemon/blob/475d7548fc16b110777f6fb6633bd09b7a7518af/internal/interfaces/grpc/handler/utils.go#L234)) which i find wrong as validation should happen in method using actual struct and not somewhere outside, since like this method is only "protected" in case it is called in this flow.
To summarise things i see problematic:
1. Representing simple structs as interfaces and then creating additional structs in different layers that implements that interface, dont see any pros only adding complexity 
2. Moving types to ports that are not "leaving" core
3. Validation structs outside of method using actual struct, not protecting method using validated struct if called from somewhere else, eg. test

Please share your thoughts.
@tiero @altafan

**Comment by tiero** [comment]
> 2\. Moving types to ports that are not "leaving" core

I agree on this, better to have little code replicated, but at least we can swap all layers without breaking anything in the future.

> Validation structs outside of method using actual struct

+1

File: resources/compose/docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    depends_on:
+      - oceand-db
+    environment:
+      - OCEAN_LOG_LEVEL=5
+      - OCEAN_NO_TLS=true
+      - OCEAN_NO_PROFILER=true
+      - OCEAN_NODE_RPC_ADDR=http://admin1:123@liquid:18884
+      - OCEAN_NETWORK=regtest
+      - OCEAN_ESPLORA_URL=http://chopsticks-liquid:3000
+      - OCEAN_UTXO_EXPIRY_DURATION_IN_SECONDS=60
```
[the remark this answers is not in the document]
Comment (altafan) [reply resources/compose/docker-compose.yml:16]: Done.

**Comment by altafan** [comment]
>     * Getting `ListTrades failure - RpcError: Skip must be set to a positive number` if `page: { number: 0, size: 5 }`. Number should be at least 1. What should be the behaviour of `number: 0`? Error message should be more explicit.

@Janaka-Steph fixed.

File: cmd/tdexd/main.go [hunk 14]
```diff
@@ -19,6 +19,7 @@ import (
 	"github.com/tdex-network/tdex-daemon/internal/core/ports"
 	oceanwallet "github.com/tdex-network/tdex-daemon/internal/infrastructure/ocean-wallet"
 	webhookpubsub "github.com/tdex-network/tdex-daemon/internal/infrastructure/pubsub/webhook"
+	swap_parser "github.com/tdex-network/tdex-daemon/internal/infrastructure/swap-parser"
```
Comment (tiero) [inline cmd/tdexd/main.go:22]:
```suggestion
	swapparser "github.com/tdex-network/tdex-daemon/internal/infrastructure/swap-parser"
```

just to be consistent?

**Comment by altafan** [comment]
> Purpose of ports pkg is to isolate core from calls to 3rd party apps/infra.
In this PR we moved almost all types to ports, even some that doesnt "leave" our core.

At the beginning, I didn't want to make the stuff defined in ports be used like cross-layer types.
You have to consider that most of the interfaces defined in ports are related to the WalletService (ocean), like for example WalletAccount, TxInput, TxOutputs, etc.

Even if we get back to the prev solution (ie. each layer has its own types, even if exactly the same), most of those interface will stay there anyway because they're related to an external service.

Then, it turned out that most of these interfaces are very similar to the proto types used by the interfaces layer, while the app layer have to deal with them in any case since it has to interact with the external wallet. Because of these reasons, I decided to use the interfaces as cross-layer types.

The main advantage of this approach is that we reduce the amount of repeated code, therefore the redundancy of potential bugs, and also allows us to write validators only once for any of those types.

> Validation structs outside of method using actual struct, not protecting method using validated struct if called from somewhere else, eg. test

I definitely agree with this and I'm glad it came out because I just forgot about it. As I said just above, it would be enough to add validators (and related unit tests) and use them in both interface and application layers where the types are used.
